new AWS KMS module

tedder opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
`kms`

##### ANSIBLE VERSION
```
ansible 2.3.0

```

##### SUMMARY
- adds/removes user/role from a KMS key. Nothing else. Good start for a KMS administration module. Needed because it can't really be done in Cloudformation.

tedder committed:
new KMS module

- add/remove user/role from a KMS key.

tedder committed:
fix spelling

tedder commented:
Need a rerun on Shippable.

ryansb commented on lib/ansible/modules/cloud/amazon/kms.py:
```diff
@@ -0,0 +1,272 @@
+
+DOCUMENTATION = '''
+---
+module: kms
```
Can you rename this as `aws_kms` please? Three-letter acronyms have enough risk for overlap that it's probably worth doing.

tedder replied:
done. it's the only `aws_` module, fwiw. But I agree on the likely overlap.

ryansb commented on lib/ansible/modules/cloud/amazon/kms.py:
```diff
@@ -0,0 +1,272 @@
+DOCUMENTATION = '''
+---
+module: kms
+short_description: Perform various KMS management tasks.
```
Maybe `Manage role/user access to a KMS key` would be more descriptive.

ryansb commented on lib/ansible/modules/cloud/amazon/kms.py:
```diff
@@ -0,0 +1,272 @@
+    required: false
+  role_name:
+    description:
+    - Role to allow/deny access. One of C(role_name) or C(role_arn) are required.
```
Please also add `role_arn` as a doc param.

ryansb commented on lib/ansible/modules/cloud/amazon/kms.py:
```diff
@@ -0,0 +1,272 @@
+    if policy['Version'] != "2012-10-17":
+        errors.append('Unknown version/date ({}) of policy. Things are probably different than we assumed they were.'.format(policy['Version']))
+
+    haz = {}
```
Not everyone on the codebase is aware that has==haz, or even has English-as-a-first-language. Can you rename this to be clearer?

ryansb commented on lib/ansible/modules/cloud/amazon/kms.py:
```diff
@@ -0,0 +1,272 @@
+
+        except Exception as err:
+            error_msg = boto_exception(err)
+            module.fail_json(msg=error_msg, traceback=traceback.format_exc().splitlines())
```
When using fail_json, please include `exception=traceback.format_exc()` as a kwarg.

ryansb commented on lib/ansible/modules/cloud/amazon/kms.py:
```diff
@@ -0,0 +1,272 @@
+        kms = ansible.module_utils.ec2.boto3_conn(module, conn_type='client', resource='kms', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+        iam = ansible.module_utils.ec2.boto3_conn(module, conn_type='client', resource='iam', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+    except botocore.exceptions.NoCredentialsError as e:
+        module.fail_json(msg=str(e))
```
When using fail_json, please include `exception=traceback.format_exc()` as a kwarg.

tedder replied:
okay. Looks like usage of this for the "create boto conn" exception is all over the map, fwiw.,

tedder committed:
move kms to aws_kms for uniqueness

tedder committed:
updates per @ryansb code review

tedder committed:
adds "invalid ARN cleanup"

If the KMS world gets orphaned ARNs (pointing to instances that no longer exist, for instance), a policy can't be edited. The cleanup is basically garbage collection; I know no reason to disable it, but leaving it there.

tedder commented:
Changes are in. Added "arn cleanup" from my own testing.

ryansb commented on lib/ansible/modules/cloud/amazon/aws_kms.py:
```diff
@@ -0,0 +1,299 @@
+                # we're granting and we recognize this statement ID.
+
+                if granttype in granttypes:
+                    if clean_invalid_entries and len(list(filter(lambda x: not x.startswith('arn:aws:iam::'), statement['Principal']['AWS']))):
```
Can you make this into a couple statements for easier readability? Something like:

```
invalid_entries = [x for x in statement['Principal']['AWS'] 
    if not x.startswith('arn:aws:iam::)]
if clean_invalid_entries and len(invalid_entries):
    for entry in invalid_entries:
        statement['Principal']['AWS'].remove(entry)
```

Or there's always:

```
original_size = len(statement['Principal']['AWS'])
statement['Principal']['AWS'] = [x for x in statement['Principal']['AWS'] 
    if x.startswith('arn:aws:iam::)]
have_invalid_entries = (original_size == len(statement['Principal']['AWS']))
```

tedder replied:
yeah, I iterated through this in a few ways. See if you like the latest.. or do you really dislike the filter()?

ryansb commented on lib/ansible/modules/cloud/amazon/aws_kms.py:
```diff
@@ -0,0 +1,299 @@
+  type: dict
+  returned: always
+  sample: { "role": "add", "role grant": "add" }
+have_invalid_entries:
```
`s/have/had/`

ryansb commented on lib/ansible/modules/cloud/amazon/aws_kms.py:
```diff
@@ -0,0 +1,299 @@
+        module.fail_json(msg='cannot connect to AWS', exception=traceback.format_exc(e))
+
+
+    if mode == 'grant' or mode == 'deny':
```
You don't need this conditional, since Ansible enforces that these are the only choices.

ryansb commented on lib/ansible/modules/cloud/amazon/aws_kms.py:
```diff
@@ -0,0 +1,299 @@
+            if module.params['key_alias'] and not module.params['key_arn']:
+                module.params['key_arn'] = get_arn_from_kms_alias(kms, module.params['key_alias'])
+            if not module.params['key_arn']:
+                module.fail_json(msg='KMS ARN is required to {}'.format(mode))
```
Add here that the `key_alias` or `key_arn` are both ways to provide it.

ryansb commented on lib/ansible/modules/cloud/amazon/aws_kms.py:
```diff
@@ -0,0 +1,299 @@
+
+    try:
+        if len(changes_needed) and not dry_run:
+            #policy_json_string = json.dumps({'Policy': policy})
```
Please remove this comment since it's not needed.

ryansb reviewed (changes requested):
Should be set to go after this set of changes.

tedder committed:
help clear up ambiguity per code review

tedder committed:
rm commented-out code
